libp2p/NodeTable.cpp

```diff
@@ -401,49 +418,60 @@ void NodeTable::onPacketReceived(
         {
             case Pong::type:
             {
-                auto in = dynamic_cast<Pong const&>(*packet);
-                // whenever a pong is received, check if it's in m_evictions
-                bool found = false;
-                NodeID leastSeenID;
-                EvictionTimeout evictionEntry;
-                DEV_GUARDED(x_evictions)
-                { 
-                    auto e = m_evictions.find(in.sourceid);
-                    if (e != m_evictions.end())
-                    {
-                        if (e->second.evictedTimePoint + c_reqTimeout >=
-                            std::chrono::steady_clock::now())
-                        {
-                            found = true;
-                            leastSeenID = e->first;
-                            evictionEntry = e->second;
-                            m_evictions.erase(e);
-                        }
-                    }
+                auto pong = dynamic_cast<Pong const&>(*packet);
+                auto const sourceId = pong.sourceid;
+
+                // validate pong
+                auto const sentPing = m_sentPings.find(sourceId);
+                if (sentPing == m_sentPings.end())
+                {
+                    LOG(m_logger) << "Unsolicited PONG from " << _from.address().to_string() << ":"
+                                  << _from.port();
+                    return;
                 }
 
-                if (found)
+                if (pong.echo != sentPing->second.second)
```

Comment:
In this case we should probably also remove it from `m_sentPings`.

Reply:
They will be removed by timeout anyway, but maybe we could remove it here right away, too.